From d02480940dfaf6dc5de5fa9d1ce12069af274147 Mon Sep 17 00:00:00 2001
From: Sertonix <sertonix@posteo.net>
Date: Tue, 12 Nov 2024 21:22:15 +0100
Subject: [PATCH] update-kernel: fix handling of compressed firmware

Since the change to zstd compressed firmware[0] update-kernel missed most
of the firmware when creating the modloop. Now firmware should be included
regardless of compression.

I have tried if decompressing the firmware for the modloop improves the
size but it doesn't. The modloop was a little bit larger and a generated
iso was exactly the same size.

Fixes https://gitlab.alpinelinux.org/alpine/alpine-conf/-/issues/10594
Fixes https://gitlab.alpinelinux.org/alpine/aports/-/issues/16601
Closes https://gitlab.alpinelinux.org/alpine/aports/-/issues/16607
Closes https://gitlab.alpinelinux.org/alpine/alpine-conf/-/merge_requests/215
Closes https://gitlab.alpinelinux.org/alpine/alpine-conf/-/merge_requests/217

[0]: https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/60282
---
 update-kernel.in | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/update-kernel.in b/update-kernel.in
index 9df2cb7..8e729a8 100644
--- a/update-kernel.in
+++ b/update-kernel.in
@@ -295,12 +295,14 @@ mkdir $MODLOOP $STAGING
 cp -a $ROOTFS/lib/modules $MODLOOP
 mkdir -p $MODLOOP/modules/firmware
 find $ROOTFS/lib/modules -type f -name "*.ko*" | xargs modinfo -k $KVER -F firmware | sort -u | while read FW; do
-	for f in "$ROOTFS"/lib/firmware/$FW; do
+	for f in "$ROOTFS"/lib/firmware/$FW "$ROOTFS"/lib/firmware/$FW.xz "$ROOTFS"/lib/firmware/$FW.zst; do
 		if ! [ -e "$f" ]; then
 			continue
 		fi
 		install -pD "$f" "$MODLOOP/modules/firmware/${f#*/lib/firmware}"
 		# copy also all potentially associated files
+		f=${f%.zst}
+		f=${f%.xz}
 		for _file in "${f%.*}".*; do
 			install -pD "$_file" "$MODLOOP/modules/firmware/${_file#*/lib/firmware/}"
 		done
@@ -327,7 +329,7 @@ fi
 
 # include bluetooth firmware in modloop
 if [ -e "$ROOTFS"/lib/modules/*/kernel/drivers/bluetooth/btbcm.ko* ]; then
-	for _btfw in "$ROOTFS"/lib/firmware/brcm/*.hcd; do
+	for _btfw in "$ROOTFS"/lib/firmware/brcm/*.hcd*; do
 		install -pD "$_btfw" \
 			"$MODLOOP"/modules/firmware/brcm/"${_btfw##*/}"
 	done
-- 
2.47.0

